### ISSUE TITLE
Icinga2 api deadlock

### ISSUE THREAD

**Issue by ymartin-ovh** [Icinga/icinga2#9916]
## Describe the bug

We experience since 2.14 with icingadb (we don't have this with icinga-ido-pgsql), icinga2 api deadlock.
The issue is visible only on Icinga master zone on node 1 (which handle all config changes). Second master is not impacted by this.

## To Reproduce

We don't know how to reproduce this but I managed to take a backtrace of all threads.
[backtrace-dbg.log](https://github.com/Icinga/icinga2/files/13452039/backtrace-dbg.log)

## Expected behavior

We expect a working Icinga2 api.

## Your Environment

Include as many relevant details about the environment you experienced the problem in

* Version used (`icinga2 --version`): 2.14.0
* Operating System and version: Debian 11
* Enabled features (`icinga2 feature list`): api checker icingadb livestatus notification perfdata
* Icinga Web 2 version and modules (System - About): 2.11.4-1
* Config validation (`icinga2 daemon -C`): OK

**Comment by ymartin-ovh** [comment]
I think that icinga::StatusHandler::HandleRequest is related to calls we have on status API (icinga2 exporter and monitoring).
I don't understand why threads are stuck on lock whereas callers give up after 5s (our configuration).

Regards

**Comment by ymartin-ovh** [comment]
Spotted deadlock between icingadb operations and add/remove comment via API:

icingadb thread =>
```
#0  __lll_lock_wait (futex=futex@entry=0x7f6c122c9c10, private=0) at lowlevellock.c:52
#1  0x00007f6ec3b938d1 in __GI___pthread_mutex_lock (mutex=0x7f6c122c9c10) at ../nptl/pthread_mutex_lock.c:115
#2  0x00005629bfeb22fa in __gthread_mutex_lock (__mutex=0x7f6c122c9c10) at /usr/include/x86_64-linux-gnu/c++/10/bits/gthr-default.h:749
#3  __gthread_recursive_mutex_lock (__mutex=0x7f6c122c9c10) at /usr/include/x86_64-linux-gnu/c++/10/bits/gthr-default.h:811
#4  std::recursive_mutex::lock (this=0x7f6c122c9c10) at /usr/include/c++/10/mutex:106
#5  icinga::ObjectLock::Lock (this=0x7f6c1aabdf00) at ./lib/base/objectlock.cpp:32
#6  0x00005629c0230a21 in icinga::Service::GetSeverity (this=this@entry=0x7f6c122c9c00) at ./lib/icinga/service.cpp:116
#7  0x00005629c0520600 in icinga::IcingaDB::SerializeState (this=<optimized out>, checkable=...) at ./lib/icingadb/icingadb-objects.cpp:2601
#8  0x00005629c0523605 in icinga::IcingaDB::UpdateState (this=0x7f6e92096000, checkable=..., mode=icinga::IcingaDB::Full) at ./lib/icingadb/icingadb-objects.cpp:1139
#9  0x00005629c0527721 in icinga::IcingaDB::SendAddedComment (this=0x7f6e92096000, comment=...) at ./lib/icingadb/icingadb-objects.cpp:2065
#10 0x00005629c0527880 in icinga::IcingaDB::CommentAddedHandler (comment=...) at ./lib/icingadb/icingadb-objects.cpp:2813
#11 0x00005629c02ffa39 in boost::function1<void, boost::intrusive_ptr<icinga::Comment> const&>::operator() (a0=..., this=<optimized out>) at /usr/include/boost/function/function_template.hpp:763
#12 boost::signals2::detail::call_with_tuple_args<boost::signals2::detail::void_type>::m_invoke<boost::function<void (boost::intrusive_ptr<icinga::Comment> const&)>, 0u, boost::intrusive_ptr<icinga::Comment> co>
    at /usr/include/boost/signals2/detail/variadic_slot_invoker.hpp:105
#13 boost::signals2::detail::call_with_tuple_args<boost::signals2::detail::void_type>::operator()<boost::function<void (boost::intrusive_ptr<icinga::Comment> const&)>, boost::intrusive_ptr<icinga::Comment> cons>
    func=..., this=<optimized out>) at /usr/include/boost/signals2/detail/variadic_slot_invoker.hpp:90
#14 boost::signals2::detail::variadic_slot_invoker<boost::signals2::detail::void_type, boost::intrusive_ptr<icinga::Comment> const&>::operator()<boost::shared_ptr<boost::signals2::detail::connection_body<std::p>
    at /usr/include/boost/signals2/detail/variadic_slot_invoker.hpp:133
#15 boost::signals2::detail::slot_call_iterator_t<boost::signals2::detail::variadic_slot_invoker<boost::signals2::detail::void_type, boost::intrusive_ptr<icinga::Comment> const&>, std::_List_iterator<boost::sha>
    at /usr/include/boost/signals2/detail/slot_call_iterator.hpp:110
#16 boost::iterators::iterator_core_access::dereference<boost::signals2::detail::slot_call_iterator_t<boost::signals2::detail::variadic_slot_invoker<boost::signals2::detail::void_type, boost::intrusive_ptr<icin>
    at /usr/include/boost/iterator/iterator_facade.hpp:550
#17 boost::iterators::detail::iterator_facade_base<boost::signals2::detail::slot_call_iterator_t<boost::signals2::detail::variadic_slot_invoker<boost::signals2::detail::void_type, boost::intrusive_ptr<icinga::C>
    at /usr/include/boost/iterator/iterator_facade.hpp:656
#18 boost::signals2::optional_last_value<void>::operator()<boost::signals2::detail::slot_call_iterator_t<boost::signals2::detail::variadic_slot_invoker<boost::signals2::detail::void_type, boost::intrusive_ptr<i>
ive_ptr<icinga::Comment> const&), boost::function<void (boost::intrusive_ptr<icinga::Comment> const&)> >, boost::signals2::mutex> >) const (last=..., this=<optimized out>, last=..., first=...) at /usr/include/b>
#19 boost::signals2::detail::combiner_invoker<void>::operator()<boost::signals2::optional_last_value<void>, boost::signals2::detail::slot_call_iterator_t<boost::signals2::detail::variadic_slot_invoker<boost::si>
#20 boost::signals2::detail::signal_impl<void (boost::intrusive_ptr<icinga::Comment> const&), boost::signals2::optional_last_value<void>, int, std::less<int>, boost::function<void (boost::intrusive_ptr<icinga::>
#21 0x00005629c0237d05 in boost::signals2::signal<void (boost::intrusive_ptr<icinga::Comment> const&), boost::signals2::optional_last_value<void>, int, std::less<int>, boost::function<void (boost::intrusive_ptr>
#22 icinga::Comment::Start (this=0x7f6c11baf600, runtimeCreated=<optimized out>) at ./lib/icinga/comment.cpp:101
#23 0x00005629bfebfc4c in icinga::ConfigObject::Activate (this=this@entry=0x7f6c11baf600, runtimeCreated=runtimeCreated@entry=true, cookie=...) at ./lib/base/configobject.cpp:368
#24 0x00005629bffe3d79 in icinga::ConfigItem::ActivateItems (newItems=std::vector of length 1, capacity 1 = {...}, runtimeCreated=runtimeCreated@entry=true, mainConfigActivation=mainConfigActivation@entry=false>
#25 0x00005629c0027516 in icinga::ConfigObjectUtility::CreateObject (type=..., fullName=..., config=..., errors=..., diagnosticInformation=..., cookie=...) at ./lib/remote/configobjectutility.cpp:248
#26 0x00005629c01c1713 in icinga::Comment::AddComment (checkable=..., entryType=<optimized out>, author=..., text=..., persistent=<optimized out>, expireTime=6.9219723785798194e-310, sticky=false, id=..., origi>
#27 0x00005629c01c216d in icinga::ApiActions::AddComment (object=..., params=...) at ./lib/icinga/apiactions.cpp:321
#28 0x00005629c027f90a in std::__invoke_impl<boost::intrusive_ptr<icinga::Dictionary>, boost::intrusive_ptr<icinga::Dictionary> (*&)(boost::intrusive_ptr<icinga::ConfigObject> const&, boost::intrusive_ptr<icing>
#29 std::__invoke_r<icinga::Value, boost::intrusive_ptr<icinga::Dictionary> (*&)(boost::intrusive_ptr<icinga::ConfigObject> const&, boost::intrusive_ptr<icinga::Dictionary> const&), boost::intrusive_ptr<icinga:>
#30 std::_Function_handler<icinga::Value (boost::intrusive_ptr<icinga::ConfigObject> const&, boost::intrusive_ptr<icinga::Dictionary> const&), boost::intrusive_ptr<icinga::Dictionary> (*)(boost::intrusive_ptr<i>
#31 0x00005629c004590e in std::function<icinga::Value (boost::intrusive_ptr<icinga::ConfigObject> const&, boost::intrusive_ptr<icinga::Dictionary> const&)>::operator()(boost::intrusive_ptr<icinga::ConfigObject>>
#32 icinga::ApiAction::Invoke (params=..., target=..., this=<optimized out>) at ./lib/remote/apiaction.cpp:14
#33 icinga::ActionsHandler::HandleRequest (this=<optimized out>, stream=..., user=..., request=..., url=..., response=..., params=..., yc=..., server=...) at ./lib/remote/actionshandler.cpp:93
#34 0x00005629c003ba04 in icinga::HttpHandler::ProcessRequest (stream=..., user=..., request=..., response=..., yc=..., server=...) at ./lib/remote/httphandler.cpp:111
#35 0x00005629c0071781 in ProcessRequest (yc=..., hasStartedStreaming=@0x7f6ce83447b1: false, server=..., response=..., authenticatedUser=..., request=..., stream=...) at ./lib/remote/httpserverconnection.cpp:4>
#36 icinga::HttpServerConnection::ProcessMessages (this=0x7f6ce8344720, yc=...) at ./lib/remote/httpserverconnection.cpp:571
#37 0x00005629c0072b20 in operator() (__closure=0x7f6c1aac1680, __closure=0x7f6c1aac1680, yc=...) at ./lib/remote/httpserverconnection.cpp:66
#38 operator() (this=0x7f6c1aac1680, yc=...) at ./lib/base/io-engine.hpp:106
#39 0x00005629c0072c6f in boost::asio::detail::coro_entry_point<boost::asio::executor_binder<void (*)(), boost::asio::io_context::strand>, icinga::IoEngine::SpawnCoroutine<boost::asio::io_context::strand, icing>
#40 boost::coroutines::detail::push_coroutine_object<boost::coroutines::pull_coroutine<void>, void, boost::asio::detail::coro_entry_point<boost::asio::executor_binder<void (*)(), boost::asio::io_context::strand>
#41 boost::coroutines::detail::trampoline_push_void<boost::coroutines::detail::push_coroutine_object<boost::coroutines::pull_coroutine<void>, void, boost::asio::detail::coro_entry_point<boost::asio::executor_bi>
#42 0x00007f6ec368619f in make_fcontext () from /lib/x86_64-linux-gnu/libboost_context.so.1.74.0
#43 0x00005629c093ca88 in vtable for boost::coroutines::detail::push_coroutine_object<boost::coroutines::pull_coroutine<void>, void, boost::asio::detail::coro_entry_point<boost::asio::executor_binder<void (*)()>
#44 0x0000000000000026 in ?? ()
#45 0x0000000000000000 in ?? ()
```
=> (gdb) print(this->GetName()) (on frame 6)
```
$20 = {static NPos = 18446744073709551615, m_Data = "cmsorderprivateusdev-002b-gra-vps.static.ovh.net!check_webovhstatic_website_dir"}
```

addcomment thread =>
```
#0  __lll_lock_wait (
    futex=futex@entry=0x5629c09746a0 <icinga::ConfigItem::ActivateItems(std::vector<boost::intrusive_ptr<icinga::ConfigItem>, std::allocator<boost::intrusive_ptr<icinga::ConfigItem> > > const&, bool, bool, bool, icinga::Value const&)::mtx>, private=0) at lowlevellock.c:52
#1  0x00007f6ec3b93843 in __GI___pthread_mutex_lock (
    mutex=0x5629c09746a0 <icinga::ConfigItem::ActivateItems(std::vector<boost::intrusive_ptr<icinga::ConfigItem>, std::allocator<boost::intrusive_ptr<icinga::ConfigItem> > > const&, bool, bool, bool, icinga::Value const&)::mtx>) at ../nptl/pthread_mutex_lock.c:80
#2  0x00005629bffe383d in __gthread_mutex_lock (
    __mutex=0x5629c09746a0 <icinga::ConfigItem::ActivateItems(std::vector<boost::intrusive_ptr<icinga::ConfigItem>, std::allocator<boost::intrusive_ptr<icinga::ConfigItem> > > const&, bool, bool, bool, icinga::Value const&)::mtx>) at /usr/include/x86_64-linux-gnu/c++/10/bits/gthr-default.h:749
#3  std::mutex::lock (
    this=0x5629c09746a0 <icinga::ConfigItem::ActivateItems(std::vector<boost::intrusive_ptr<icinga::ConfigItem>, std::allocator<boost::intrusive_ptr<icinga::ConfigItem> > > const&, bool, bool, bool, icinga::Value const&)::mtx>) at /usr/include/c++/10/bits/std_mutex.h:100
#4  std::unique_lock<std::mutex>::lock (this=0x7f6c36c6b820) at /usr/include/c++/10/bits/unique_lock.h:138
#5  std::unique_lock<std::mutex>::unique_lock (__m=..., this=0x7f6c36c6b820) at /usr/include/c++/10/bits/unique_lock.h:68
#6  icinga::ConfigItem::ActivateItems (newItems=std::vector of length 1, capacity 1 = {...}, runtimeCreated=runtimeCreated@entry=true, mainConfigActivation=mainConfigActivation@entry=false, 
    withModAttrs=withModAttrs@entry=false, cookie=...) at ./lib/config/configitem.cpp:683
#7  0x00005629c0027516 in icinga::ConfigObjectUtility::CreateObject (type=..., fullName=..., config=..., errors=..., diagnosticInformation=..., cookie=...) at ./lib/remote/configobjectutility.cpp:248
#8  0x00005629c01c1713 in icinga::Comment::AddComment (checkable=..., entryType=<optimized out>, author=..., text=..., persistent=<optimized out>, expireTime=6.9219956746066321e-310, sticky=true, id=..., 
    origin=...) at ./lib/icinga/comment.cpp:170
#9  0x00005629c022c0a3 in icinga::ApiActions::AcknowledgeProblem (object=..., params=...) at ./lib/icinga/apiactions.cpp:265
#10 0x00005629c027f90a in std::__invoke_impl<boost::intrusive_ptr<icinga::Dictionary>, boost::intrusive_ptr<icinga::Dictionary> (*&)(boost::intrusive_ptr<icinga::ConfigObject> const&, boost::intrusive_ptr<icinga::Dictionary> const&), boost::intrusive_ptr<icinga::ConfigObject> const&, boost::intrusive_ptr<icinga::Dictionary> const&> (__f=<optimized out>) at /usr/include/c++/10/bits/invoke.h:59
#11 std::__invoke_r<icinga::Value, boost::intrusive_ptr<icinga::Dictionary> (*&)(boost::intrusive_ptr<icinga::ConfigObject> const&, boost::intrusive_ptr<icinga::Dictionary> const&), boost::intrusive_ptr<icinga::ConfigObject> const&, boost::intrusive_ptr<icinga::Dictionary> const&> (__fn=<optimized out>, __fn=<optimized out>) at /usr/include/c++/10/bits/invoke.h:113
#12 std::_Function_handler<icinga::Value (boost::intrusive_ptr<icinga::ConfigObject> const&, boost::intrusive_ptr<icinga::Dictionary> const&), boost::intrusive_ptr<icinga::Dictionary> (*)(boost::intrusive_ptr<icinga::ConfigObject> const&, boost::intrusive_ptr<icinga::Dictionary> const&)>::_M_invoke(std::_Any_data const&, boost::intrusive_ptr<icinga::ConfigObject> const&, boost::intrusive_ptr<icinga::Dictionary> const&) (__functor=..., __args#0=..., __args#1=...) at /usr/include/c++/10/bits/std_function.h:292
#13 0x00005629c004590e in std::function<icinga::Value (boost::intrusive_ptr<icinga::ConfigObject> const&, boost::intrusive_ptr<icinga::Dictionary> const&)>::operator()(boost::intrusive_ptr<icinga::ConfigObject> const&, boost::intrusive_ptr<icinga::Dictionary> const&) const (__args#1=..., __args#0=..., this=<optimized out>) at /usr/include/c++/10/bits/std_function.h:622
#14 icinga::ApiAction::Invoke (params=..., target=..., this=<optimized out>) at ./lib/remote/apiaction.cpp:14
#15 icinga::ActionsHandler::HandleRequest (this=<optimized out>, stream=..., user=..., request=..., url=..., response=..., params=..., yc=..., server=...) at ./lib/remote/actionshandler.cpp:93
--Type <RET> for more, q to quit, c to continue without paging--
#16 0x00005629c003ba04 in icinga::HttpHandler::ProcessRequest (stream=..., user=..., request=..., response=..., yc=..., server=...) at ./lib/remote/httphandler.cpp:111
#17 0x00005629c0071781 in ProcessRequest (yc=..., hasStartedStreaming=@0x7f6ce3efdd31: false, server=..., response=..., authenticatedUser=..., request=..., stream=...)
    at ./lib/remote/httpserverconnection.cpp:456
#18 icinga::HttpServerConnection::ProcessMessages (this=0x7f6ce3efdca0, yc=...) at ./lib/remote/httpserverconnection.cpp:571
#19 0x00005629c0072b20 in operator() (__closure=0x7f6c36c6dd00, __closure=0x7f6c36c6dd00, yc=...) at ./lib/remote/httpserverconnection.cpp:66
#20 operator() (this=0x7f6c36c6dd00, yc=...) at ./lib/base/io-engine.hpp:106
#21 0x00005629c0072c6f in boost::asio::detail::coro_entry_point<boost::asio::executor_binder<void (*)(), boost::asio::io_context::strand>, icinga::IoEngine::SpawnCoroutine<boost::asio::io_context::strand, icinga::HttpServerConnection::Start()::<lambda(boost::asio::yield_context)> >::<lambda(boost::asio::yield_context)> >::operator() (ca=..., this=<optimized out>) at /usr/include/boost/asio/impl/spawn.hpp:355
#22 boost::coroutines::detail::push_coroutine_object<boost::coroutines::pull_coroutine<void>, void, boost::asio::detail::coro_entry_point<boost::asio::executor_binder<void (*)(), boost::asio::io_context::strand>, icinga::IoEngine::SpawnCoroutine<boost::asio::io_context::strand, icinga::HttpServerConnection::Start()::<lambda(boost::asio::yield_context)> >::<lambda(boost::asio::yield_context)> >&, boost::coroutines::basic_standard_stack_allocator<boost::coroutines::stack_traits> >::run (this=0x7f6c36c6e060) at /usr/include/boost/coroutine/detail/push_coroutine_object.hpp:302
#23 boost::coroutines::detail::trampoline_push_void<boost::coroutines::detail::push_coroutine_object<boost::coroutines::pull_coroutine<void>, void, boost::asio::detail::coro_entry_point<boost::asio::executor_binder<void (*)(), boost::asio::io_context::strand>, icinga::IoEngine::SpawnCoroutine<boost::asio::io_context::strand, icinga::HttpServerConnection::Start()::<lambda(boost::asio::yield_context)> >::<lambda(boost::asio::yield_context)> >&, boost::coroutines::basic_standard_stack_allocator<boost::coroutines::stack_traits> > >(boost::context::detail::transfer_t) (t=...)
    at /usr/include/boost/coroutine/detail/trampoline_push.hpp:70
#24 0x00007f6ec368619f in make_fcontext () from /lib/x86_64-linux-gnu/libboost_context.so.1.74.0
#25 0x00005629c093ca88 in vtable for boost::coroutines::detail::push_coroutine_object<boost::coroutines::pull_coroutine<void>, void, boost::asio::detail::coro_entry_point<boost::asio::executor_binder<void (*)(), boost::asio::io_context::strand>, icinga::IoEngine::SpawnCoroutine<boost::asio::io_context::strand, icinga::HttpServerConnection::Start()::{lambda(boost::asio::basic_yield_context<boost::asio::executor_binder<void (*)(), boost::asio::executor> >)#1}>(boost::asio::io_context::strand&, icinga::HttpServerConnection::Start()::{lambda(boost::asio::basic_yield_context<boost::asio::executor_binder<void (*)(), boost::asio::executor> >)#1})::{lambda(boost::asio::basic_yield_context<boost::asio::executor_binder<void (*)(), boost::asio::executor> >)#1}>&, boost::coroutines::basic_standard_stack_allocator<boost::coroutines::stack_traits> > ()
#26 0x0000000000000026 in ?? ()
#27 0x0000000000000000 in ?? ()
```

=> info args on frame 7:
```
type = @0x7f6c36c6c160: {px = 0x5629c0947168 <vtable for std::basic_ofstream<char, std::char_traits<char> >+24>}
fullName = @0x7f6c36c6cb90: {static NPos = 18446744073709551615, m_Data = "cmsorderprivateusdev-002b-gra-vps.static.ovh.net!check_webovhstatic_website_dir!3345cdff-f2b0-476c-827e-2326f20e40b9"}
config = @0x7f6c36c6bfb0: {static NPos = 18446744073709551615, m_Data = "\000\000\000\000\000\000\000\000\000J\364\300n\177\000\000\220!a\347l\177\000\000\240!a\347l\177\000\000\240"}
errors = @0x7f6c36c6c360: {px = 0x7f6c00000357}
diagnosticInformation = @0x7f6c36c6c780: {px = 0x0}
cookie = @0x5629c09729a0: {m_Value = {which_ = 0, storage_ = {<boost::detail::aligned_storage::aligned_storage_imp<32, 8>> = {data_ = {buf = '\000' <repeats 31 times>, align_ = {<No data fields>}}}, 
      static size = <optimized out>, static alignment = <optimized out>}}}
```

The first thread on this report took the static std::mutex in activateitems and wait for object lock whereas the second done the opposite.

**Comment by ymartin-ovh** [comment]
Checking with Icinga2 object lock debug:

On icingadb thread:
```
(gdb) print(this->m_LockOwner)
$21 = {static _S_min_alignment = 8, static _S_alignment = 8, _M_i = {_M_thread = 140112867874560}, static is_always_lock_free = <optimized out>}
```
140112867874560 => 7F6E91B7D700 is thread 57 (the one with addcomment)

**Comment by ymartin-ovh** [comment]
Hello @Al2Klimov  / @julianbrost 

Did you encounter this ?

Regards

**Comment by ymartin-ovh** [comment]
Hello

I tested your proposal and I still experience deadlock on Icinga2 API with this.

Regards

**Comment by ymartin-ovh** [comment]
Hi @yhabteab 

Thanks for the feedback and no problem for the delay !

Regards

**Comment by julianbrost** [comment]
> However, if you now release the `olock` prematurely, another `/v1/actions/acknowledge-problem` request might get the chance to proceed and mistakenly assume that the checkable isn't acked, while in fact there's another ongoing acknowledgement request.

So your suggestion would be to leave that `ObjectLock` untouched there, which means the call to `Comment::AddComment()` would eventually end up in `ConfigItem::ActivateItems()` with the `ObjectLock` still held with the risk of deadlocking with the mutex there?

https://github.com/Icinga/icinga2/blob/d230e87cf2167aa8f74a4d1af297a9ed82972109/lib/config/configitem.cpp#L642-L646

> I would rather question the need for the static mutex in the ConfigItem::ActivateItems() method in the first place. It's introduced with [050c520](https://github.com/Icinga/icinga2/commit/050c520b2a3df0477f2eec71ec2d874b6dd36aa5), but I don't know the exact reasoning behind it. Looking at the code, I don't even see why we would want to lock any mutex there apart from maybe the `if (withModAttrs)` part.

It's pretty clear what that mutex achieves: it prevents the whole function from ever running concurrently. Ignoring the `if (withModAttrs)` part for the moment: what exactly is the claim you're making here? Are you saying that the function is safe to run concurrently without any restrictions? Or are there other mechanism already in place that would already restrict it to safe situations?

Regarding the `if (withModAttrs)` part: The only call to `ActivateItems()` with `withModAttrs = true` seems to be this one:
https://github.com/Icinga/icinga2/blob/d230e87cf2167aa8f74a4d1af297a9ed82972109/lib/cli/daemoncommand.cpp#L302

That's only done once during startup, so unless that's changes, that block specifically shouldn't need any extra protection.

**Comment by yhabteab** [comment]
> So your suggestion would be to leave that `ObjectLock` untouched there, which means the call to `Comment::AddComment()` would eventually end up in `ConfigItem::ActivateItems()` with the `ObjectLock` still held with the risk of deadlocking with the mutex there?

I'm not sure, I get this right but are you aware of where exactly the deadlock occurs or which mutexes are involved? Yes, I mean, that we can leave the existing olock in `ApiActions::AcknowledgeProblem()` untouched if we get the static mutex from `ActivateItems()` dropped, otherwise we would have to introduce yet a new mutex only for the acknowledgement part.

> It's pretty clear what that mutex achieves: it prevents the whole function from ever running concurrently.
> what exactly is the claim you're making here? Are you saying that the function is safe to run concurrently without any restrictions?

I'm saying that there's literally nothing that should be protected within that method in any case. Yes, it does prevent the whole function from running concurrently now, but what's the reason for that? Which resource does it protect specifically? Because I don't see any. Hence my suggestion to drop it completely if there's no specific reason to not to.

**Comment by julianbrost** [comment]
> I'm not sure, I get this right but are you aware of where exactly the deadlock occurs or which mutexes are involved?

Well the point of my question was to figure that out as I haven't this in great detail yet. Not sure if it was unclear which mutexes I was referring to in my question. So are these two the ones involved?

https://github.com/Icinga/icinga2/blob/1a386ad55d3049a14d863dcff536f6817aafbba0/lib/icinga/apiactions.cpp#L240
https://github.com/Icinga/icinga2/blob/1a386ad55d3049a14d863dcff536f6817aafbba0/lib/config/configitem.cpp#L646

The PR suggests to reduce the scope of the first one, you see a problem and suggest to remove the other one entirely.

> Yes, it does prevent the whole function from running concurrently now, but what's the reason for that? Which resource does it protect specifically? Because I don't see any.

I don't see something obvious so far either, it should also have the effect that it prevents all `ConfigObject::Start()` implementations from running concurrently. I'm not yet sure if something could be hidden there.

**Comment by yhabteab** [comment]
> Not sure if it was unclear which mutexes I was referring to in my question. So are these two the ones involved?

Yes, both are involved while the letter can also be acquired from e.g. the following actions handler without acquiring the first one:
https://github.com/Icinga/icinga2/blob/abd95a33ca074399d17ed842c8a75419db16a1ba/lib/icinga/apiactions.cpp#L296

The deadlock in this specific case occurs as follows:
1. Thread **A** processes a `/v1/actions/acknowledge-problem` request and locks the [checkable](https://github.com/Icinga/icinga2/blob/master/lib/icinga/apiactions.cpp#L240)
2. Thread **B** processes a [`/v1/actions/add-comment`](https://github.com/Icinga/icinga2/blob/master/lib/icinga/apiactions.cpp#L296) and enters first the [`ConfigItem::ActivateItems()`](https://github.com/Icinga/icinga2/blob/master/lib/config/configitem.cpp#L642-L646) method and locks the static mutex there and starts the just created comment object, which triggers the [`OnCommentAdded()`](https://github.com/Icinga/icinga2/blob/master/lib/icinga/comment.cpp#L100-L101) event.
3. Thread **A** wants to activate the just created ack comment as well but since the [mutex](https://github.com/Icinga/icinga2/blob/master/lib/config/configitem.cpp#L646) is already locked by TA it blocks.
4. Thread **B**'s `OnCommentAdded()` even dispatch causes the `IcingaDB::CommentAddedHandler()` to be called and implicitly triggers full state update for the [`checkable`](https://github.com/Icinga/icinga2/blob/master/lib/icingadb/icingadb-objects.cpp#L2334). Now, the state serialization of that checkable (remember that's the same checkable currently locked by TA) also includes computing its severity, thus it calls either [`service->GetSeverity()`](https://github.com/Icinga/icinga2/blob/master/lib/icingadb/icingadb-objects.cpp#L2962) or [`host->GetSeverity()`](https://github.com/Icinga/icinga2/blob/master/lib/icingadb/icingadb-objects.cpp#L2969). However, since computing the checkable severity (as of now) requires acquiring the object lock, and boom - they deadlock each other.

> The PR suggests to reduce the scope of the first one, you see a problem and suggest to remove the other one entirely.

Yes, by removing the mutex from `ActivateItems()` the deadlock would be automatically fixed without touching the olock in the API actions.

> it should also have the effect that it prevents all `ConfigObject::Start()` implementations from running concurrently. I'm not yet sure if something could be hidden there.

I doubt that the mutex was introduced because of the `Start()` method, but I can't prove it either, so I'll look into it tomorrow.

**Comment by julianbrost** [comment]
After having looked at this a few more times, I fail to find a reason for that mutex either:

https://github.com/Icinga/icinga2/blob/d230e87cf2167aa8f74a4d1af297a9ed82972109/lib/config/configitem.cpp#L645-L646

As already said, the only really interesting task it performs is (indirectly) calling `Start()` on the individual objects, but historically, they were called in parallel and also when looking at the individual `Start()`, they make use of synchronization mechanisms (like `boost::call_once`), i.e. expect to be called in parallel. So if there was a reason for that mutex, only the Gods of Icinga know and if we remove it, worst case we'll figure out why it was there. But as multiple developers now failed to see a reason, I doubt there is one and vote for removing it.

**Comment by yhabteab** [comment]
@ymartin-ovh can you please remove the mutex listed in https://github.com/Icinga/icinga2/issues/9916#issuecomment-2897620461 in your PR and revert all the other changes you already made. As no one from our side is able to tell what that mutex is used for, you can just drop it and run some tests on your setup to verify whether it fixes your specific issue and share your findings! Thanks!

**Comment by ymartin-ovh** [comment]
Hello

I tested using curl doing parallel add-comment/remove-comment and acknowledge-problem/remove-acknowledgement on a single service.
Deadlock seems resolved but I trigger a new issue. I can trigger this issue too if I push hard on add-comment/remove-comment.
Maybe is a new bug.

```
Stacktrace:
 0# icinga::Application::SigAbrtHandler(int) in /usr/lib/x86_64-linux-gnu/icinga2/sbin/icinga2
 1# 0x00007F7EE17F8140 in /lib/x86_64-linux-gnu/libpthread.so.0
 2# gsignal in /lib/x86_64-linux-gnu/libc.so.6
 3# abort in /lib/x86_64-linux-gnu/libc.so.6
 4# 0x00007F7EE130840F in /lib/x86_64-linux-gnu/libc.so.6
 5# 0x00007F7EE13176E2 in /lib/x86_64-linux-gnu/libc.so.6
 6# 0x000055D4F5CF98F3 in /usr/lib/x86_64-linux-gnu/icinga2/sbin/icinga2
 7# icinga::ApiActions::AddComment(boost::intrusive_ptr<icinga::ConfigObject> const&, boost::intrusive_ptr<icinga::Dictionary> const&) in /usr/lib/x86_64-linux-gnu/icinga2/sbin/icinga2
 8# std::_Function_handler<icinga::Value (boost::intrusive_ptr<icinga::ConfigObject> const&, boost::intrusive_ptr<icinga::Dictionary> const&), boost::intrusive_ptr<icinga::Dictionary> (*)(boost::intrusive_ptr<icinga::ConfigObject> const&, boost::intrusive_ptr<icinga::Dictionary> const&)>::_M_invoke(std::_Any_data const&, boost::intrusive_ptr<icinga::ConfigObject> const&, boost::intrusive_ptr<icinga::Dictionary> const&) in /usr/lib/x86_64-linux-gnu/icinga2/sbin/icinga2
 9# icinga::ActionsHandler::HandleRequest(icinga::AsioTlsStream&, boost::intrusive_ptr<icinga::ApiUser> const&, boost::beast::http::message<true, boost::beast::http::basic_string_body<char, std::char_traits<char>, std::allocator<char> >, boost::beast::http::basic_fields<std::allocator<char> > >&, boost::intrusive_ptr<icinga::Url> const&, boost::beast::http::message<false, boost::beast::http::basic_string_body<char, std::char_traits<char>, std::allocator<char> >, boost::beast::http::basic_fields<std::allocator<char> > >&, boost::intrusive_ptr<icinga::Dictionary> const&, boost::asio::basic_yield_context<boost::asio::executor_binder<void (*)(), boost::asio::executor> >&, icinga::HttpServerConnection&) in /usr/lib/x86_64-linux-gnu/icinga2/sbin/icinga2
10# icinga::HttpHandler::ProcessRequest(icinga::AsioTlsStream&, boost::intrusive_ptr<icinga::ApiUser> const&, boost::beast::http::message<true, boost::beast::http::basic_string_body<char, std::char_traits<char>, std::allocator<char> >, boost::beast::http::basic_fields<std::allocator<char> > >&, boost::beast::http::message<false, boost::beast::http::basic_string_body<char, std::char_traits<char>, std::allocator<char> >, boost::beast::http::basic_fields<std::allocator<char> > >&, boost::asio::basic_yield_context<boost::asio::executor_binder<void (*)(), boost::asio::executor> >&, icinga::HttpServerConnection&) in /usr/lib/x86_64-linux-gnu/icinga2/sbin/icinga2
11# icinga::HttpServerConnection::ProcessMessages(boost::asio::basic_yield_context<boost::asio::executor_binder<void (*)(), boost::asio::executor> >) in /usr/lib/x86_64-linux-gnu/icinga2/sbin/icinga2
12# 0x000055D4F5C01460 in /usr/lib/x86_64-linux-gnu/icinga2/sbin/icinga2
13# 0x000055D4F5C015AF in /usr/lib/x86_64-linux-gnu/icinga2/sbin/icinga2
14# make_fcontext in /lib/x86_64-linux-gnu/libboost_context.so.1.74.0
```


With gdb stacktrace where I can find the abort call:
```
#5  0x00007f7ee1308537 in abort () from /lib/x86_64-linux-gnu/libc.so.6
No symbol table info available.
#6  0x00007f7ee130840f in ?? () from /lib/x86_64-linux-gnu/libc.so.6
No symbol table info available.
#7  0x00007f7ee13176e2 in __assert_fail () from /lib/x86_64-linux-gnu/libc.so.6
No symbol table info available.
#8  0x000055d4f5cf98f3 in boost::intrusive_ptr<icinga::Comment>::operator-> (this=<optimized out>) at /usr/include/boost/smart_ptr/intrusive_ptr.hpp:199
        __PRETTY_FUNCTION__ = {<optimized out> <repeats 74 times>}
#9  0x000055d4f5d4dc25 in boost::intrusive_ptr<icinga::Comment>::operator-> (this=<synthetic pointer>) at ./lib/icinga/apiactions.cpp:333
        __PRETTY_FUNCTION__ = {<optimized out> <repeats 74 times>}
#10 icinga::ApiActions::AddComment (object=..., params=...) at ./lib/icinga/apiactions.cpp:327
        checkable = {px = 0x7f7df49c4a00}
        timestamp = <optimized out>
        lock = {m_Lock = {mp_mutex = 0x7f7ee1e86000, m_locked = true}}
        commentName = {static NPos = 18446744073709551615, m_Data = "ymartin.sdev-docker.ha.ovh.net!check_disk!a350e20d-1235-4216-8a83-183c2459aa58"}
        comment = {px = <optimized out>}
        additional = {px = 0x7f7c23850010}
```

**Comment by yhabteab** [comment]
> Deadlock seems resolved but I trigger a new issue. I can trigger this issue too if I push hard on add-comment/remove-comment.
> Maybe is a new bug.

Indeed ;). You just unveiled yet another Icinga 2 crash. Though, it has nothing to do with the changes you made in your PR. That's just yet another hidden bug that was waiting to be discovered. If you're willing to, you can easily fix it in your PR though. The bug only get triggered when you remove the just created comment fast enough, before the creating comment request isn't done yet.

You create the comment here...
https://github.com/Icinga/icinga2/blob/3d04eb456add0294347ff0148f3af316564086b6/lib/icinga/apiactions.cpp#L318-L320

...and definitely pass through this check, otherwise we'd have seen a `Could not create comment` error:
https://github.com/Icinga/icinga2/blob/3d04eb456add0294347ff0148f3af316564086b6/lib/icinga/comment.cpp#L179-L182

And right after this you remove that comment using the `/v1/actions/remove-comment` endpoint, after which this look up yield no comment but a nullptr.
https://github.com/Icinga/icinga2/blob/3d04eb456add0294347ff0148f3af316564086b6/lib/icinga/apiactions.cpp#L322

Finally, when calling `comment->GetLegacyId()` on that nullptr here, it ultimately crashes with nullptr dereference.
https://github.com/Icinga/icinga2/blob/3d04eb456add0294347ff0148f3af316564086b6/lib/icinga/apiactions.cpp#L326

The easy fix for this is to just let `Comment::AddComment()` return the comment pointer instead of the string name, and the lookup by name in `ApiActions::AddComment()` won't be needed anymore.

**Comment by ymartin-ovh** [comment]
Thanks for the analysis @yhabteab . I added a commit on my PR. I saw a similar work on downtime done by @julianbrost in 2021.
I'm currently packaging icinga2 with this 2 commits and will test them.

Regards

**Comment by ymartin-ovh** [comment]
Hello

Tested on a particular service in critical state with:
* 40 concurrents acknowledge-problem remove-acknowledgement
* 20 concurrents add-comment remove-comment

Seems good (150000 requests done).
